src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -633,17 +724,56 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
             return BindParameterFromValue(parameter, Expression.Coalesce(routeValue, queryValue), factoryContext);
         }
 
-        private static Expression BindParameterFromBody(Type parameterType, bool allowEmpty, FactoryContext factoryContext)
+        private static Expression BindParameterFromBody(ParameterInfo parameter, bool allowEmpty, FactoryContext factoryContext)
         {
             if (factoryContext.JsonRequestBodyType is not null)
             {
                 throw new InvalidOperationException("Action cannot have more than one FromBody attribute.");
             }
 
-            factoryContext.JsonRequestBodyType = parameterType;
-            factoryContext.AllowEmptyRequestBody = allowEmpty;
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            factoryContext.JsonRequestBodyType = parameter.ParameterType;
+            factoryContext.AllowEmptyRequestBody = allowEmpty || isOptional;
+
+            var argument = Expression.Variable(parameter.ParameterType, $"{parameter.Name}_local");
```

Comment:
Why do we need this? Can't we just null-check `BodyValueExpr`?

Reply:
We need this because the way the try-parse logic is generated. From what I've been able to reason from removing this argument and tracing through the code, the try-parse logic assembles a code block that looks like this when mapped to a lambda.

```
(target, httpContext, bodyValue) => {
  (param1_local, param2_local) {
    // if statements for validation here
   // check tryParseFailedValueHere
   // call action here
  }
}
```

If we try to do something like:

```
(target, httpContext, bodyValue) => {
  (param1_local, bodyValue) {
    // if statements for validation here
   // check tryParseFailedValueHere
   // call action here
  }
}
```

Then the original value of `bodyValue` is shadowed and set to null so the null-check doesn't behave as we'd expect.

Reply:
The try-parse logic shouldn't touch `bodyValue` since the body is never TryParsed.

Reply:
> try-parse logic 

try-parse here refers to the expression blocks that are added to the `TryParseParams` list and executed. Without this argument the `checkBodyParamRequired` expression would've shadowed bodyValue.

Reply:
I'm a bit confused here too. The TryParse should never matter for the body argument.

Reply:
If you want to check if the body is null after casting, and don't want to cast twice, storing the casted value in the local helps. Not sure if that's necessary, but one extra local doesn't seem that bad.

Reply:
> I'm a bit confused here too. The TryParse should never matter for the body argument.

Previously, it didn't. This PR updates the `TryParse` logic to be a generic codepath for validating parameters so now we are using it for body arguments.

> If you want to check if the body is null after casting, and don't want to cast twice, storing the casted value in the local helps. Not sure if that's necessary, but one extra local doesn't seem that bad.

Yeah, it's not the worse. Can you validate my understanding of what is happening mentioned in https://github.com/dotnet/aspnetcore/pull/34505#discussion_r673498034?